Std/Data/Array/Basic.lean

```diff
@@ -117,6 +117,69 @@ protected def maxI [ord : Ord α] [Inhabited α]
     (xs : Array α) (start := 0) (stop := xs.size) : α :=
   xs.minI (ord := ord.opposite) start stop
 
+/--
+Returns the first and all the rest of the elements of the array.
+-/
+protected def splitFront (xs : Array α) (h : 0 < xs.size) : α × Array α :=
+  let head := xs.get ⟨0, h⟩
+  let tail := Subarray.mk xs 1 xs.size (Nat.succ_le_of_lt h) (Nat.le_refl _) |>.toArray
+  (head, tail)
+
+/--
+Returns the first and all the rest of the elements of the array, or `none` if it is empty.
+-/
+protected def splitFront? (xs : Array α) : Option (α × Array α) :=
+  if h : 0 < xs.size then
+    some <| xs.splitFront h
+  else
+    none
+
+/--
+Returns the last and all the rest of the elements of the array.
+-/
+protected def splitLast (xs : Array α) (h : 0 < xs.size) : Array α × α :=
+  have h2 := Nat.sub_lt h (by decide)
+  let last := xs.get ⟨xs.size - 1, h2⟩
+  let heads := Subarray.mk xs 0 (xs.size - 1) (Nat.zero_le _) (Nat.le_of_lt h2) |>.toArray
```

Comment:
This should use `pop` or its underlying machinery, because this will avoid making a copy if the input array is unshared